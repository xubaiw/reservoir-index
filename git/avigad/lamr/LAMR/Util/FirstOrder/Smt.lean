import Init
import Lean.Parser
import Lean.PrettyPrinter

section
open Lean.Parser
open Lean.PrettyPrinter

/-- Like `ident` but with no splitting on dots and accepts anything that's not whitespace
or parentheses. So e.g. `<=` works. -/
def generalIdent : Parser :=
  withAntiquot (mkAntiquot "generalIdent" `generalIdent) {
    fn := fun c s =>
      let startPos := s.pos
      let s := takeWhile1Fn (fun c => !("(){}[].".contains c) ∧ !c.isWhitespace) "expected generalized identifier" c s
      mkNodeToken `generalIdent startPos c s }

def Lean.TSyntax.getGeneralId : TSyntax `generalIdent → String
  | ⟨Syntax.node _ `generalIdent args⟩ => args[0]!.getAtomVal!
  | s => panic! s!"unexpected syntax '{s}'"

@[combinatorFormatter generalIdent] def generalIdent.formatter : Formatter := pure ()
@[combinatorParenthesizer generalIdent] def generalIdent.parenthesizer : Parenthesizer := pure ()
end

inductive Sexp where
  | atom : String → Sexp
  | expr : List Sexp → Sexp
  deriving Repr, BEq, Inhabited

namespace Sexp

instance : Coe String Sexp :=
  ⟨Sexp.atom⟩

declare_syntax_cat sexp

syntax generalIdent : sexp
syntax "(" sexp* ")" : sexp
syntax "(" sexp* "...{" term "}" sexp* ")" : sexp
syntax "{" term "}" : sexp

-- This coercion is justified by the macro expansions below.
instance : Coe (Lean.TSyntax `sexp) (Lean.TSyntax `term) where
  coe a := ⟨a.raw⟩

macro_rules
  | `(sexp| $a:generalIdent) => `(Sexp.atom $(Lean.quote a.getGeneralId))
  | `(sexp| ( $ss:sexp* )) => `(Sexp.expr [ $ss,* ])
  | `(sexp| ( $ss:sexp* ...{ $t:term } $ts:sexp* )) => `(Sexp.expr <| [ $ss,* ] ++ ($t : List Sexp) ++ [ $ts,* ])
  | `(sexp| { $t:term }) => `($t)

syntax "sexp!{" sexp "}" : term
macro_rules
  | `(sexp!{ $s:sexp }) => `($s)

syntax "sexps!{" sexp* "}" : term
syntax "sexps!{" sexp* "...{" term "}" sexp* "}" : term
macro_rules
  | `(sexps!{ $ss:sexp* }) => do
    let ss ← ss.mapM fun s => `(sexp!{ $s })
    `([ $[$ss],* ])
  | `(sexps!{ $ss:sexp* ...{ $t:term } $ts:sexp* }) =>
    `([ $[$ss],* ] ++ ($t : List Sexp) ++ [ $[$ts],* ])

partial def serialize : Sexp → String
  | atom s  => s
  | expr ss => s!"({" ".intercalate <| ss.map serialize})"

partial def serializeMany (ss : List Sexp) : String :=
  ss.map serialize |> "\n".intercalate

instance : ToString Sexp :=
  ⟨serialize⟩

instance : Repr Sexp where
  reprPrec s _ := s!"sexp!\{{toString s}}"

partial def parse (s : String) : Except String (List Sexp) :=
  let tks := tokenize #[] s.toSubstring
  parseMany #[] tks.toList |>.map Prod.fst |>.map Array.toList
where
  tokenize (stk : Array Substring) (s : Substring) : Array Substring :=
    if s.isEmpty then stk
    else
      let c := s.front
      if c == ')' || c == '(' then
        tokenize (stk.push <| s.take 1) (s.drop 1)
      else if c.isWhitespace then tokenize stk (s.drop 1)
      else
        let tk := s.takeWhile fun c => !c.isWhitespace && c != '(' && c != ')'
        if tk.bsize > 0 then tokenize (stk.push tk) (s.extract ⟨tk.bsize⟩ ⟨s.bsize⟩)
        else unreachable!

  parseOne : List Substring → Except String (Sexp × List Substring)
    | tk :: tks => do
      if tk.front == ')' then
        throw "mismatched parentheses"
      if tk.front == '(' then
        let (ss, tks) ← parseMany #[] tks
        return (expr ss.toList, tks)
      else
        return  (atom tk.toString, tks)
    | [] => throw "expected input, got none"

  parseMany (stk : Array Sexp) : List Substring → Except String (Array Sexp × List Substring)
    | tk :: tks => do
      if tk.front == ')' then .ok (stk, tks)
      else
        let (e, tks) ← parseOne (tk :: tks)
        parseMany (stk.push e) tks
    | [] => .ok (stk, [])

end Sexp

private def argsCvc4 : IO.Process.SpawnArgs := {
  cmd := "LAMR/bin/cvc4"
  args := #["--lang", "smt", "LAMR/bin/temp.smt"] }

private def argsCvc5 : IO.Process.SpawnArgs := {
  cmd := "LAMR/bin/cvc5"
  args := #["--lang", "smt", "LAMR/bin/temp.smt"] }

private def argsZ3 : IO.Process.SpawnArgs := {
  cmd := "LAMR/bin/z3"
  args := #["-smt2", "LAMR/bin/temp.smt"] }

private def argsBoolector : IO.Process.SpawnArgs := {
  cmd := "LAMR/bin/boolector"
  args := #["--smt2", "LAMR/bin/temp.smt"] }

-- Same as IO.Process.run, but does not require exitcode = 0
private def run' (args : IO.Process.SpawnArgs) : IO String := do
  let out ← IO.Process.output args
  pure out.stdout

/-- Executes the solver with the provided list of commands in SMT-LIB s-expression format.
Returns the solver output as s-expressions. -/
private def callSolver (args : IO.Process.SpawnArgs) (commands : List Sexp) (verbose : Bool := false)
    : IO (List Sexp) := do
  let cmdStr := Sexp.serializeMany commands
  if verbose then
    IO.println "Sending SMT-LIB problem:"
    IO.println cmdStr
  IO.FS.writeFile "LAMR/bin/temp.smt" cmdStr
  let out ← run' args
  if verbose then
    IO.println "\nSolver replied:"
    IO.println out
  let out ← IO.ofExcept (Sexp.parse out)
  return out

def callCvc4 := @callSolver argsCvc4
def callCvc5 := @callSolver argsCvc5
def callZ3 := @callSolver argsZ3
def callBoolector := @callSolver argsBoolector

private def hexdigits : Array Char :=
  #[ '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f' ]

private def enhexByte (x : UInt8) : String :=
  ⟨[hexdigits.get! $ UInt8.toNat $ (x.land 0xf0).shiftRight 4,
    hexdigits.get! $ UInt8.toNat $ x.land 0xf ]⟩

/-- Convert a little-endian (LSB first) list of bytes to hexadecimal. -/
private def enhexLE : List UInt8 → String
  | [] => ""
  | b::bs => enhexLE bs ++ enhexByte b

/-- Converts a number `n` to its hexadecimal SMT-LIB representation as a `nBits`-bit vector.
For example `toBVConst 32 0xf == "#x0000000f"`. -/
def toBVConst (nBits : Nat) (n : Nat) : String :=
  assert! nBits % 8 == 0
  let nBytes := nBits/8
  let bytes := List.range nBytes |>.map fun i => UInt8.ofNat ((n >>> (i*8)) &&& 0xff)
  "#x" ++ enhexLE bytes

open Std (AssocList)

/-- Extracts constants assigned in a model returned from an SMT solver.
The model is expected to be a single s-expression representing a list,
with constant expressions represented by `(define-fun <name> () <type> <body>)`. -/
def decodeModelConsts : Sexp → AssocList String Sexp
  | Sexp.expr ss =>
    ss.foldl (init := AssocList.empty) fun
      | acc, sexp!{(define-fun {Sexp.atom x} () {_} {body})} =>
        acc.insert x body
      | acc, _ => acc
  | _ => AssocList.empty

/-- Evaluates an SMT-LIB constant numeral such as `0` or `#b01` or `#x02`. -/
def evalNumConst : Sexp → Option Nat
  | Sexp.atom s =>
    let s' :=
      if s.startsWith "#b" then "0" ++ s.drop 1
      else if s.startsWith "#x" then "0" ++ s.drop 1
      else s
    Lean.Syntax.decodeNatLitVal? s'
  | Sexp.expr _ => none