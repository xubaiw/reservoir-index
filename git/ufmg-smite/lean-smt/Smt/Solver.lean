/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean
import Smt.Term

namespace Smt

inductive Kind where
  | cvc4
  | cvc5
  | z3
  | yices
  | boolector
  | vampire
  deriving Inhabited, DecidableEq

instance : ToString Kind where
  toString
    | .cvc4      => "cvc4"
    | .cvc5      => "cvc5"
    | .z3        => "z3"
    | .yices     => "yices"
    | .boolector => "boolector"
    | .vampire   => "vampire"

instance : Lean.KVMap.Value Kind where
  toDataValue k := toString k
  ofDataValue?
    | "cvc4"      => Kind.cvc4
    | "cvc5"      => Kind.cvc5
    | "z3"        => Kind.z3
    | "yices"     => Kind.yices
    | "boolector" => Kind.boolector
    | "vampire"   => Kind.vampire
    | _           => none

/-- What the binary for a given solver is usually called. -/
def Kind.toDefaultPath : Kind → String
  | .yices => "yices-smt2"
  | k      => toString k

register_option smt.solver.kind : Kind := {
  defValue := Kind.cvc5
  descr := "The solver to use for solving the SMT query."
}

register_option smt.solver.path : String := {
  defValue := "cvc5"
  descr := "The path to the solver used for solving the SMT query."
}

/-- The data-structure for SMT-LIB based solver. -/
structure Solver where
  commands : List String
  deriving Inhabited

namespace Solver
open Term

variable (solver : Solver)

/-- Set the SMT query logic to `l`. -/
def setLogic (l : String) : Solver :=
  ⟨s!"(set-logic {l})" :: solver.commands⟩

/-- Define a sort with name `id` and arity `n`. -/
def declareSort (id : String) (n : Nat) : Solver :=
  ⟨s!"(declare-sort {quoteSymbol id} {n})" :: solver.commands⟩

/-- Define a sort with name `id`. -/
def defineSort (id : String) (ss : List Term) (s : Term) : Solver :=
  ⟨(s!"(define-sort {quoteSymbol id} ({paramsToString ss}) {s})") :: solver.commands⟩
  where
    paramsToString (ss) := String.intercalate " " (ss.map Term.toString)

/-- Declare a symbolic constant with name `id` and sort `s`. -/
def declareConst (id : String) (s : Term) : Solver :=
  ⟨s!"(declare-const {quoteSymbol id} {s})" :: solver.commands⟩

/-- Define a constant with name `id` sort `s`, and value `v`. -/
def defineConst (id : String) (s : Term) (v : Term) : Solver :=
  ⟨s!"(define-const {quoteSymbol id} {s} {v})" :: solver.commands⟩

/-- Declare an uninterpreted function with name `id` and sort `s`. -/
def declareFun (id : String) (s : Term) : Solver :=
  ⟨s!"(declare-fun {quoteSymbol id} {s})" :: solver.commands⟩

/-- Define a function with name `id`, parameters `ps`, co-domain `s`,
    and body `t`. -/
def defineFun (id : String) (ps : List (String × Term)) (s : Term) (t : Term) :
  Solver :=
  ⟨s!"(define-fun {quoteSymbol id} ({paramsToString ps}) {s} {t})" :: solver.commands⟩
  where
    paramsToString (ps : List (String × Term)) : String :=
      String.intercalate " " (ps.map (fun (n, s) => s!"({quoteSymbol n} {s})"))

/-- Define a recursive function with name `id`, parameters `ps`, co-domain `s`,
    and body `t`. -/
def defineFunRec (id : String) (ps : List (String × Term)) (s : Term)
  (t : Term) :
  Solver :=
  ⟨s!"(define-fun-rec {quoteSymbol id} ({paramsToString ps}) {s} {t})" :: solver.commands⟩
  where
    paramsToString (ps : List (String × Term)) : String :=
      String.intercalate " " (ps.map (fun (n, s) => s!"({quoteSymbol n} {s})"))

/-- Assert that proposition `t` is true. -/
def assert (t : Term) : Solver :=
  ⟨s!"(assert {t})" :: solver.commands⟩

def queryToString : String :=
  String.intercalate "\n" ("(check-sat)\n" :: solver.commands).reverse

/-- Check if the query given so far is satisfiable and return the result. -/
def checkSat (kind : Kind) (path : String) (timeoutSecs? : Option Nat := some 5) : IO String := do
  let mut args := match kind with
    | .cvc4      => #["--quiet", "--lang", "smt"]
    | .cvc5      => #["--quiet", "--lang", "smt"]
    | .z3        => #["-in", "-smt2"]
    | .yices     => #[]
    | .boolector => #["--smt2"]
    | .vampire   => #["--input_syntax", "smtlib2", "--output_mode", "smtcomp"]
  if let some secs := timeoutSecs? then
    args := args ++ match kind with
    | .cvc4      => #["--tlimit", toString (secs*1000)]
    | .cvc5      => #["--tlimit", toString (secs*1000)]
    | .z3        => #[s!"-T:{secs}"]
    | .yices     => #["--timeout", toString secs]
    | .boolector => #["--time", toString secs]
    | .vampire   => #["--time_limit", toString secs]
  let proc ← IO.Process.spawn {
    stdin := IO.Process.Stdio.piped
    stdout := IO.Process.Stdio.piped
    stderr := IO.Process.Stdio.piped
    cmd := path
    args
  }
  let query := solver.queryToString ++ "\n(exit)"
  proc.stdin.putStr query
  -- Close stdin to signal EOF to the solver.
  let (_, proc) ← proc.takeStdin
  _ ← proc.wait
  let out ← proc.stdout.readToEnd
  let err ← proc.stderr.readToEnd
  return s!"{out}{if err.isEmpty then "" else "\n"}{err}"

end Solver
end Smt
