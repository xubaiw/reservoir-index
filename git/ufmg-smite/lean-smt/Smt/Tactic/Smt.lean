/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean

import Smt.Dsl.Sexp
import Smt.Query
import Smt.Solver

namespace Smt

open Lean
open Lean.Elab
open Lean.Elab.Tactic

initialize
  registerTraceClass `smt.debug
  registerTraceClass `smt.debug.attr
  registerTraceClass `smt.debug.query
  registerTraceClass `smt.debug.translator

syntax smtHints := ("[" ident,* "]")?
syntax smtTimeout := ("(timeout := " num ")")?

/-- `smt` converts the current goal into an SMT query and checks if it is
satisfiable. By default, `smt` generates the minimum valid SMT query needed to
assert the current goal. However, that is not always enough:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt
```
For the theorem above, `smt` generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(check-sat)
```
which is missing the hypotheses `hp` and `hpq` required to prove the theorem. To
pass hypotheses to the solver, use `smt [h₁, h₂, ..., hₙ]` syntax:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt [hp, hpq]
```
The tactic then generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(declare-const p Bool)
(assert p)
(assert (=> p q))
(check-sat)
```
-/
syntax (name := smt) "smt" smtHints smtTimeout : tactic

/-- Like `smt`, but just shows the query without invoking a solver. -/
syntax (name := smtShow) "smt_show" smtHints : tactic

def parseHints : TSyntax `smtHints → TacticM (List Expr)
  | `(smtHints| [ $[$hs],* ]) => hs.toList.mapM (fun h => elabTerm h.raw none)
  | `(smtHints| ) => return []
  | _ => throwUnsupportedSyntax

def prepareSmtQuery (hints : TSyntax `smtHints) : TacticM Solver := do
  -- 1. Get the current main goal.
  let goalType ← Tactic.getMainTarget
  let goalId ← Lean.mkFreshMVarId
  Lean.Meta.withLocalDeclD goalId.name (mkNot goalType) fun g => do
  -- 2. Get the hints passed to the tactic.
  let mut hs ← parseHints hints
  hs := hs.eraseDups
  -- 3. Generate the SMT query.
  let mut solver := Solver.mk []
  Query.generateQuery g hs solver

@[tactic smt] def evalSmt : Tactic := fun stx => withMainContext do
  let goalType ← Tactic.getMainTarget
  let solver ← prepareSmtQuery ⟨stx[1]⟩
  let query := solver.queryToString
  -- 4. Run the solver.
  let kind := smt.solver.kind.get (← getOptions)
  let path := smt.solver.path.get? (← getOptions) |>.getD kind.toDefaultPath
  -- Don't run solver if the server cancelled our task
  if (← IO.checkCanceled) then return
  let res ← if let `(smtTimeout| (timeout := $n)) := stx[2] then
    solver.checkSat kind path (some n.getNat)
  else
    solver.checkSat kind path
  -- 5. Print the result.
  logInfo m!"goal: {goalType}\n\nquery:\n{query}\nresult: {res}"
  let out ← match Sexp.parse res with
    | .ok out => pure out
    | .error e => throwError "cannot parse solver output: {e}"
  if out.contains sexp!{sat} then
    throwError "unable to prove goal, either it is false or you need to define more symbols with `smt [foo, bar]`"
    -- TODO ask for countermodel and print
  else if !out.contains sexp!{unsat} then
    throwError "unable to prove goal"

@[tactic smtShow] def evalSmtShow : Tactic := fun stx => withMainContext do
  let goalType ← Tactic.getMainTarget
  let solver ← prepareSmtQuery ⟨stx[1]⟩
  let query := solver.queryToString
  -- 4. Print the query.
  logInfo m!"goal: {goalType}\n\nquery:\n{query}"

end Smt
