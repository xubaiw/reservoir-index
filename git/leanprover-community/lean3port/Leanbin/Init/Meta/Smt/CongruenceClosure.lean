/-
Copyright (c) 2016 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
prelude
import Leanbin.Init.Meta.InteractiveBase
import Leanbin.Init.Meta.Tactic
import Leanbin.Init.Meta.SetGetOptionTactics

structure CcConfig where
  -- If tt, congruence closure will treat implicit instance arguments as constants.
  ignoreInstances : Bool := true
  -- If tt, congruence closure modulo AC.
  ac : Bool := true
  /- If ho_fns is (some fns), then full (and more expensive) support for higher-order functions is
     *only* considered for the functions in fns and local functions. The performance overhead is described in the paper
     "Congruence Closure in Intensional Type Theory". If ho_fns is none, then full support is provided
     for *all* constants. -/
  hoFns : Option (List Name) := none
  -- If true, then use excluded middle
  em : Bool := true

/-- Congruence closure state.
This may be considered to be a set of expressions and an equivalence class over this set.
The equivalence class is generated by the equational rules that are added to the cc_state and congruence,
that is, if `a = b` then `f(a) = f(b)` and so on.
 -/
unsafe axiom cc_state : Type

unsafe axiom cc_state.mk_core : CcConfig → cc_state

/-- Create a congruence closure state object using the hypotheses in the current goal. -/
unsafe axiom cc_state.mk_using_hs_core : CcConfig → tactic cc_state

/-- Get the next element in the equivalence class.
Note that if the given expr e is not in the graph then it will just return e. -/
unsafe axiom cc_state.next : cc_state → expr → expr

/-- Returns the root expression for each equivalence class in the graph.
If the bool argument is set to true then it only returns roots of non-singleton classes. -/
unsafe axiom cc_state.roots_core : cc_state → Bool → List expr

/-- Get the root representative of the given expression. -/
unsafe axiom cc_state.root : cc_state → expr → expr

/--
"Modification Time". The field m_mt is used to implement the mod-time optimization introduce by the Simplify theorem prover.
The basic idea is to introduce a counter gmt that records the number of heuristic instantiation that have
occurred in the current branch. It is incremented after each round of heuristic instantiation.
The field m_mt records the last time any proper descendant of of thie entry was involved in a merge. -/
unsafe axiom cc_state.mt : cc_state → expr → Nat

/-- "Global Modification Time". gmt is a number stored on the cc_state,
it is compared with the modification time of a cc_entry in e-matching. See `cc_state.mt`. -/
unsafe axiom cc_state.gmt : cc_state → Nat

/-- Increment the Global Modification time. -/
unsafe axiom cc_state.inc_gmt : cc_state → cc_state

/-- Check if `e` is the root of the congruence class. -/
unsafe axiom cc_state.is_cg_root : cc_state → expr → Bool

/-- Pretty print the entry associated with the given expression. -/
unsafe axiom cc_state.pp_eqc : cc_state → expr → tactic format

/-- Pretty print the entire cc graph.
If the bool argument is set to true then singleton equivalence classes will be omitted. -/
unsafe axiom cc_state.pp_core : cc_state → Bool → tactic format

/-- Add the given expression to the graph. -/
unsafe axiom cc_state.internalize : cc_state → expr → tactic cc_state

/-- Add the given proof term as a new rule.
The proof term p must be an `eq _ _`, `heq _ _`, `iff _ _`, or a negation of these. -/
unsafe axiom cc_state.add : cc_state → expr → tactic cc_state

/-- Check whether two expressions are in the same equivalence class. -/
unsafe axiom cc_state.is_eqv : cc_state → expr → expr → tactic Bool

/-- Check whether two expressions are not in the same equivalence class. -/
unsafe axiom cc_state.is_not_eqv : cc_state → expr → expr → tactic Bool

/-- Returns a proof term that the given terms are equivalent in the given cc_state-/
unsafe axiom cc_state.eqv_proof : cc_state → expr → expr → tactic expr

/-- Returns true if the cc_state is inconsistent. For example if it had both `a = b` and `a ≠ b` in it.-/
unsafe axiom cc_state.inconsistent : cc_state → Bool

/-- `proof_for cc e` constructs a proof for e if it is equivalent to true in cc_state -/
unsafe axiom cc_state.proof_for : cc_state → expr → tactic expr

/-- `refutation_for cc e` constructs a proof for `not e` if it is equivalent to false in cc_state -/
unsafe axiom cc_state.refutation_for : cc_state → expr → tactic expr

/-- If the given state is inconsistent, return a proof for false. Otherwise fail. -/
unsafe axiom cc_state.proof_for_false : cc_state → tactic expr

namespace CcState

unsafe def mk : cc_state :=
  cc_state.mk_core {  }

unsafe def mk_using_hs : tactic cc_state :=
  cc_state.mk_using_hs_core {  }

unsafe def roots (s : cc_state) : List expr :=
  cc_state.roots_core s true

unsafe instance : has_to_tactic_format cc_state :=
  ⟨fun s => cc_state.pp_core s true⟩

unsafe def eqc_of_core (s : cc_state) : expr → expr → List expr → List expr
  | e, f, r =>
    let n := s.next e
    if n = f then e :: r else eqc_of_core n f (e :: r)

unsafe def eqc_of (s : cc_state) (e : expr) : List expr :=
  s.eqc_of_core e e []

unsafe def in_singlenton_eqc (s : cc_state) (e : expr) : Bool :=
  s.next e = e

unsafe def eqc_size (s : cc_state) (e : expr) : Nat :=
  (s.eqc_of e).length

unsafe def fold_eqc_core {α} (s : cc_state) (f : α → expr → α) (first : expr) : expr → α → α
  | c, a =>
    let new_a := f a c
    let next := s.next c
    if expr.alpha_eqv next first then new_a else fold_eqc_core next new_a

unsafe def fold_eqc {α} (s : cc_state) (e : expr) (a : α) (f : α → expr → α) : α :=
  fold_eqc_core s f e e a

unsafe def mfold_eqc {α} {m : Type → Type} [Monadₓ m] (s : cc_state) (e : expr) (a : α) (f : α → expr → m α) : m α :=
  fold_eqc s e (return a) fun act e => do
    let a ← act
    f a e

end CcState

open Tactic

unsafe def tactic.cc_core (cfg : CcConfig) : tactic Unit := do
  intros
  let s ← cc_state.mk_using_hs_core cfg
  let t ← target
  let s ← s.internalize t
  if s then do
      let pr ← s
      mk_app `false.elim [t, pr] >>= exact
    else do
      let tr ← return <| expr.const `true []
      let b ← s t tr
      if b then do
          let pr ← s t tr
          mk_app `of_eq_true [pr] >>= exact
        else do
          let dbg ← get_bool_option `trace.cc.failure ff
          if dbg then do
              let ccf ← pp s
              fail
                  f! "cc tactic failed, equivalence classes: 
                    {ccf}"
            else do
              fail "cc tactic failed"

unsafe def tactic.cc : tactic Unit :=
  tactic.cc_core {  }

unsafe def tactic.cc_dbg_core (cfg : CcConfig) : tactic Unit :=
  save_options <| set_bool_option `trace.cc.failure true >> tactic.cc_core cfg

unsafe def tactic.cc_dbg : tactic Unit :=
  tactic.cc_dbg_core {  }

unsafe def tactic.ac_refl : tactic Unit := do
  let (lhs, rhs) ← target >>= match_eq
  let s ← return <| cc_state.mk
  let s ← s.internalize lhs
  let s ← s.internalize rhs
  let b ← s.is_eqv lhs rhs
  if b then do
      s lhs rhs >>= exact
    else do
      fail "ac_refl failed"

