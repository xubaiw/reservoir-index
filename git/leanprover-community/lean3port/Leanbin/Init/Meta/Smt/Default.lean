/-
Copyright (c) 2017 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
prelude
import Leanbin.Init.Meta.Smt.CongruenceClosure
import Leanbin.Init.CcLemmas
import Leanbin.Init.Meta.Smt.Ematch
import Leanbin.Init.Meta.Smt.SmtTactic
import Leanbin.Init.Meta.Smt.Interactive
import Leanbin.Init.Meta.Smt.Rsimp

