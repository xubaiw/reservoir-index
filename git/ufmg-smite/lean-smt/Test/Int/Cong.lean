import Smt

theorem cong (x y : Int) (f : Int → Int) : x = y → f x = f y := by
  smt
  simp_all
