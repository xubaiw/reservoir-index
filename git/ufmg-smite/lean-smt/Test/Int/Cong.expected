goal: x = y â†’ f x = f y

query:
(declare-const x Int)
(declare-const y Int)
(declare-fun f (Int) Int)
(assert (not (=> (= x y) (= (f x) (f y)))))
(check-sat)

result: unsat
